hydrolib/core/io/ini/serializer.py

```diff
@@ -0,0 +1,246 @@
+            )
+        else:
+            max_key_length = 0
+            max_value_length = 0
+
+        return cls(
+            max_key_length=max_key_length,
+            max_value_length=max_value_length,
+            max_datablock_element_length=_Lengths._calculate_max_length_datablock(
+                section.datablock
+            ),
+        )
+
+
+class Serializer:
+    """Serializer serializes Document to its corresponding lines."""
+
+    def __init__(self, config: SerializerConfig):
+        """Creates a new Serializer with the provided configuration.
+
+        Args:
+            config (SerializerConfig): The configuration of this Serializer.
+        """
+        self._config = config
+
+    @staticmethod
+    def _interweave(iterable: Iterable, val: Any) -> Iterable:
+        # Interweave the provided iterable with the provided value:
+        # iterable_element, val, iterable_element, val, ...
+
+        # Note that this will interweave with val without making copies
+        # as such it is the same object being interweaved.
+        return chain.from_iterable(zip(iterable, repeat(val)))
+
+    def _serialize_comment_block(
+        self, block: CommentBlock, indent_size: int = 0
+    ) -> Iterable[str]:
+        indent = " " * indent_size
+        return (f"{indent}{self._config.comment_delimiter} {l}" for l in block.lines)
+
+    def _serialize_comment_header(
+        self, header_blocks: Iterable[CommentBlock]
+    ) -> Iterable[str]:
+        blocks_as_lines = (
+            self._serialize_comment_block(block) for block in header_blocks
+        )
+        # We separate comment blocks in the header with an empty line.
+        blocks_as_lines = Serializer._interweave(blocks_as_lines, [""])  # type: ignore
+
+        return chain.from_iterable(blocks_as_lines)
+
+    def _serialize_section_header(self, section_header: str) -> Iterable[str]:
+        indent = " " * (self._config.section_indent)
+        return [f"{indent}[{section_header}]"]
+
+    @staticmethod
+    def _get_offset(key: Optional[str], max_length: int) -> str:
+        key_length = len(key) if key is not None else 0
+        return " " * max(max_length - key_length, 0)
+
+    def _serialize_property(
+        self, property: Property, lengths: _Lengths
+    ) -> Iterable[str]:
+        indent = " " * (self._config.total_property_indent)
+        key_offset = Serializer._get_offset(property.key, lengths.max_key_length)
+        key = f"{property.key}{key_offset} = "
+
+        value_offset = Serializer._get_offset(property.value, lengths.max_value_length)
+        value = (
+            f"{property.value}{value_offset}"
+            if property.value is not None
+            else value_offset
+        )
+
+        comment = f" # {property.comment}" if property.comment is not None else ""
+
+        return [f"{indent}{key}{value}{comment}".rstrip()]
+
+    def _serialize_content(
```

Comment:
Can there be a comment block between to sections? Is it then "absorbed" by the upper section?

Reply:
Currently this is not possible, every comment block is assumed to be part of the section before it, except for the first comment blocks, which are part of the document header. Unfortunately it is not really possible to make this distinction easily, therefor I made this assumption as it was easiest to implement.